File: backend-legacy-python/app/vpn/links.py
"""Subscription link generation — iterates protocol registry to build client links."""

import base64
import datetime
import logging

import redis.asyncio as aioredis

from app.vpn.protocols import registry
from app.vpn.protocols.base import ClientLink, ServerConfig, UserCredentials

logger = logging.getLogger(__name__)


def generate_all_links(
    user: UserCredentials,
    servers: list[ServerConfig],
    relay_servers: list[dict] | None = None,
) -> list[ClientLink]:
    """Generate all client links by iterating enabled protocols."""
    links: list[ClientLink] = []
    for protocol in registry.with_links():
        try:
            links.extend(protocol.client_links(user, servers))
        except Exception as e:
            logger.warning("Link generation failed for %s: %s", protocol.name, e)
    return links


async def generate_all_links_async(
    user: UserCredentials,
    servers: list[ServerConfig],
    relay_servers: list[dict] | None = None,
    check_health: bool = True,
    redis: aioredis.Redis | None = None,
) -> list[ClientLink]:
    """Generate links, optionally filtering out unhealthy servers."""
    if check_health and redis:
        healthy_servers = await _get_healthy_servers(servers, redis)
    else:
        healthy_servers = servers

    links: list[ClientLink] = []
    for protocol in registry.with_links():
        try:
            links.extend(protocol.client_links(user, healthy_servers))
        except Exception as e:
            logger.warning("Link generation failed for %s: %s", protocol.name, e)
    return links


async def _get_healthy_servers(
    servers: list[ServerConfig], redis: aioredis.Redis,
) -> list[ServerConfig]:
    """Filter servers by health status from Redis.

    Reads ``node_health:{ip}`` keys written by proxy_monitor.
    If no health data exists for a server, it is assumed healthy
    so that subscriptions are never empty.
    """
    healthy: list[ServerConfig] = []
    for srv in servers:
        key = f"node_health:{srv.host}"
        status = await redis.get(key)
        if status is None or status == b"up":
            healthy.append(srv)
        else:
            logger.debug("Excluding unhealthy server %s (%s)", srv.name, srv.host)

    # Never return empty — fall back to all servers
    return healthy if healthy else servers


def format_subscription_text(
    links: list[ClientLink],
    expire_ts: int | None = None,
    branding: dict | None = None,
) -> str:
    """Format subscription response: header lines + VPN link URIs."""
    b = branding or {}
    brand = b.get("brand_name", "Chameleon VPN")
    support = b.get("support_contact", "")
    channel = b.get("channel_handle", "")

    if expire_ts:
        date_str = datetime.datetime.fromtimestamp(expire_ts, tz=datetime.timezone.utc).strftime("%d.%m.%Y")
        header = f"{brand} | До: {date_str}"
    else:
        header = f"{brand} | Без ограничений"

    lines = [header, f"Поддержка: {support}", f"Канал: {channel}", ""]
    lines.extend(link.uri for link in links)
    return "\n".join(lines)


def get_subscription_headers(
    expire_ts: int | None = None,
    upload: int = 0,
    download: int = 0,
    branding: dict | None = None,
) -> dict:
    """Generate subscription response HTTP headers."""
    b = branding or {}
    return {
        "Cache-Control": "no-cache",
        "profile-title": base64.b64encode(b.get("profile_title", "Chameleon VPN").encode()).decode(),
        "profile-update-interval": b.get("update_interval", "12"),
        "support-url": b.get("support_url", ""),
        "profile-web-page-url": b.get("web_page_url", ""),
        "Subscription-Userinfo": f"upload={upload}; download={download}; total=0; expire={expire_ts or 0}",
    }

File: backend/crates/chameleon-vpn/src/links.rs
//! Subscription link generation — iterates protocol registry to build client links.

use base64::Engine;
use chrono::{DateTime, Utc, TimeZone};
use tracing::warn;

use super::protocols::{ClientLink, ServerConfig, UserCredentials, ProtocolRegistry};

/// Generate all client links by iterating enabled protocols.
pub fn generate_all_links(
    registry: &ProtocolRegistry,
    user: &UserCredentials,
    servers: &[ServerConfig],
) -> Vec<ClientLink> {
    let mut links = Vec::new();
    for protocol in registry.enabled() {
        links.extend(protocol.client_links(user, servers));
    }
    links
}

/// Format subscription response: only VPN link URIs, one per line.
/// Metadata (brand, expiry, support) is sent via HTTP headers instead.
/// This format is compatible with v2rayN, Streisand, Hiddify, and other clients.
pub fn format_subscription_text(
    links: &[ClientLink],
    _expire_ts: Option<i64>,
    _branding: Option<&BrandingInfo>,
) -> String {
    links.iter().map(|l| l.uri.clone()).collect::<Vec<_>>().join("\n")
}

/// Generate subscription response HTTP headers.
pub fn get_subscription_headers(
    expire_ts: Option<i64>,
    upload: i64,
    download: i64,
    branding: Option<&BrandingInfo>,
) -> Vec<(String, String)> {
    let b = branding.cloned().unwrap_or_else(|| BrandingInfo::default());
    let b64 = base64::engine::general_purpose::STANDARD;
    vec![
        ("Cache-Control".into(), "no-cache".into()),
        ("profile-title".into(), b64.encode(b.profile_title.as_bytes())),
        ("profile-update-interval".into(), b.update_interval.clone()),
        ("support-url".into(), b.support_url.clone()),
        ("profile-web-page-url".into(), b.web_page_url.clone()),
        ("Subscription-Userinfo".into(), format!("upload={upload}; download={download}; total=0; expire={}", expire_ts.unwrap_or(0))),
    ]
}

#[derive(Debug, Clone)]
pub struct BrandingInfo {
    pub brand_name: String,
    pub profile_title: String,
    pub support_contact: String,
    pub channel_handle: String,
    pub support_url: String,
    pub web_page_url: String,
    pub update_interval: String,
}

impl Default for BrandingInfo {
    fn default() -> Self {
        Self {
            brand_name: "Chameleon VPN".into(),
            profile_title: "Chameleon VPN".into(),
            support_contact: String::new(),
            channel_handle: String::new(),
            support_url: String::new(),
            web_page_url: String::new(),
            update_interval: "12".into(),
        }
    }
}
